Allow editing of base_url and handle API errors / exceptions

EdwardAndress opened the pull request:
- Added edit + create actions to admin/environments_controller
- Added edit template
- Updated tests for admin/environments_controller
- Ensure client_keys and JWT do not change on update
- Added a new ApiResponse class to abstract error handling (tests still required for this)

@aldavidson - I'm not quite happy with the new changes the ApiClient but am not entirely sure how to improve it so I've opened this up for your comments.  I'll be away now until Wednesday afternoon unfortunately.

EdwardAndress committed:
Environment#base_url will change; allow editing

EdwardAndress committed:
Env client_keys & jwt mustn't change on update

EdwardAndress committed:
ApiResponse class for better error handling

aldavidson commented:
I'm getting an error running rake db:setup, looks like it's during loading the seeds - 

```
-- initialize_schema_migrations_table()
   -> 0.0006s
rake aborted!
ActiveRecord::RecordInvalid: Validation failed: Base url can't be blank
(...)
/Users/alistairdavidson/vendor/bundle/gems/activesupport-5.0.2/lib/active_support/dependencies.rb:287:in `load'
/Users/alistairdavidson/vendor/bundle/gems/railties-5.0.2/lib/rails/engine.rb:549:in `load_seed'
/Users/alistairdavidson/vendor/bundle/gems/activerecord-5.0.2/lib/active_record/tasks/database_tasks.rb:268:in `load_seed'
(...)
Tasks: TOP => db:setup => db:seed
```

I'm guessing that base_url also needs populating in the seeds.rb?

aldavidson commented:
@EdwardAndress for the record, we talked about the "leaky abstraction" in the environment model/view/controller, and how it could be made better by making the population of the environment properties from the remote api more explicit - i.e. adding a ! method in the model to set the properties, and adding a calling method in the controller.
Thanks!

EdwardAndress committed:
Env seed data is invalid without base_url

EdwardAndress committed:
Response is exception-safe; clarify use of API

- Renamed ApiResponse to ExceptionSafeResponse to better communicate that it's purpose is to handle Exceptions
- Refactor to clarify what is going on with API calls and how #health, #deployed_version and #deployed_version_timestamp are set on an Environment

EdwardAndress commented:
@aldavidson - I am wondering whether ExceptionSafeResponse is really an ExceptionSafeResponseParser / Generator and whether I should be injecting that dependency into the NomisApiClient #new method call.

It would look something like

```ruby
# where the client is instantiated in the Environment model
def populate_properties!
    client = NomisApiClient.new(self, ExceptionSafeResponseParser.new)
    self.health = client.get_health
    self.deployed_version = client.get_version
    self.deployed_version_timestamp = client.get_version_timestamp
  end

# in NomisApiClient.rb

def initialize(env, response_parser)
  @env_name = env.name
  @client_key = env.client_private_key
  @client_token = env.jwt
  @base_url = env.base_url
  @response_parser = response_parser
end

def get(path)
    begin
      response_parser.parse(NOMIS::API::Get.new(
        client_key: client_key,
        client_token: client_token,
        base_url: base_url,
        path: path
      ).execute)
    rescue => exception
      response_parser.parse(exception)
    end
  end
```

What do you think?

aldavidson commented:
sounds reasonable

On 6 Oct 2017 16:06, "Edward Andress" <notifications@github.com> wrote:

> @aldavidson <https://github.com/aldavidson> - I am wondering whether
> ExceptionSafeResponse is really an ExceptionSafeResponseParser / Generator
> and whether I should be injecting that dependency into the NomisApiClient
> #new method call.
>
> It would look something like
>
> # where the client is instantiated in the Environment modeldef populate_properties!
>     client = NomisApiClient.new(self, ExceptionSafeResponseParser.new)
>     self.health = client.get_health
>     self.deployed_version = client.get_version
>     self.deployed_version_timestamp = client.get_version_timestamp
>   end
> # in NomisApiClient.rb
> def initialize(env, response_parser)
>   @env_name = env.name
>   @client_key = env.client_private_key
>   @client_token = env.jwt
>   @base_url = env.base_url
>   @response_parser = response_parserend
> def get(path)
>     begin
>       response_parser.parse(NOMIS::API::Get.new(
>         client_key: client_key,
>         client_token: client_token,
>         base_url: base_url,
>         path: path
>       ).execute)
>     rescue => exception
>       response_parser.parse(exception)
>     end
>   end
>
> What do you think?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ministryofjustice/noms-api-gateway-management/pull/57#issuecomment-334782175>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAINZabfYGSQOUs00bn4EMLlIncPZVT2ks5spkIFgaJpZM4PowHE>
> .
>

EdwardAndress commented:
Cool! Thanks.

EdwardAndress committed:
ExceptionSafeResponse is really a parser

- ExceptionSafeResponse renamed to ExceptionSafeResponseParser
- Injecting parser dependency into NomisApiClient at instantiation

EdwardAndress commented:
@aldavidson I think this is an improvement.  What do you think?

EdwardAndress commented on app/services/nomis_api_client.rb:
```diff
@@ -1,21 +1,18 @@
-    else
-      "ERROR #{response.raw_response.code}"
-    end
+    get('health').data
```
This was doing some sneaky parsing so it's moved into the actual parser.

EdwardAndress commented on app/services/nomis_api_client.rb:
```diff
@@ -27,11 +24,15 @@ def get_version_timestamp
+        path: path
+      ).execute)
+    rescue => exception
+      parser.parse(exception)
```
by this point we should know that it's an ExceptionSafeThing so I have used a shorter instance var

EdwardAndress commented on db/seeds.rb:
```diff
@@ -1,6 +1,19 @@
 # Development seeds
+def generate_base_url(env_name)
```
The envs created on dev, staging or pre-prod will not be able to correctly query the API so the actual base_url we're setting here doesn't matter too much for the time being, but I thought we should make it follow the same pattern anyway.  It might at some point become outdated if the url pattern changes.

EdwardAndress commented on config/routes.rb:
```diff
@@ -18,7 +18,9 @@
-    resources :environments, only: [:index, :new, :create, :show, :destroy]
+    resources :environments, only: [
+      :index, :show, :new, :create, :edit, :update, :destroy
+    ]
```
I'm mostly trying to keep lines under 80 chars, with some exceptions where readability is impacted.  This is because Misha likes lines under than length to allow 3 columns side-by-side in a text editor.  How do the rest of our team feel about this?

aldavidson commented:
I'm getting another error now:
<img width="828" alt="screen shot 2017-10-10 at 12 34 00" src="https://user-images.githubusercontent.com/134501/31384574-6e6fbe2e-adb7-11e7-9c2a-7abd2f7d41df.png">

I think there's an edge case here where the response object is not actually a response(!)
I was testing this against a dev environment on https://apgpw.al.nomis-api.hmpps.dsd.io/ - which has a self-signed SSL certificate. The 'curl' operation underneath all of the Ruby abstractions is failing because by default it doesn't trust self-signed certificates, and it's failing at SSL handshake time - _before performing the actual GET_ - so there is no response.

There's a couple of ways to handle that which I would suggest, either:
1. in the environments_controller, where it calls populate_properties! on each environment, or
2. in the populate_properties! method in the environment.rb model

In either place, you could catch the exception and probably set environment.health to "error" & print the full error details to the log.

EdwardAndress committed:
Handle SSLError in exception_safe_response_parser

EdwardAndress commented:
@aldavidson Hey Al, how about this instead?  Now we can handle two specific errors and it won't break if there's an unexpected error.

I'm going to look into Logging now.

EDIT - I just realised that I have refactored the wrong line!  On it now.

EdwardAndress commented on app/services/exception_safe_response_parser.rb:
```diff
@@ -0,0 +1,44 @@
+    data.is_a?(Hash) ? data : "#{parsed_response.code}#{': ' + data.truncate(30) if data }"
+  end
+
+  def parse_exception(response)
```
I suppose this is a little opaque because we're handling the 'non-response' Exception as if it were a real response, inside the response parser... what do you think?  Is this potentially confusing?

EdwardAndress replied:
@aldavidson  ^

EdwardAndress committed:
Exeptions raised by API calls should be logged

EdwardAndress committed:
Handle, rather than parse, Exception

aldavidson commented:
nearly there! I think you need the fully-scoped path of the SSLError:
<img width="1010" alt="screen shot 2017-10-12 at 15 55 42" src="https://user-images.githubusercontent.com/134501/31502802-fc675f84-af65-11e7-952f-30018874df48.png">

EdwardAndress committed:
Full path out of module to SSLError required

EdwardAndress commented:
Are you OK with the non-standard code that I've been inspired to use (it is used by Cloudflare)

"525 SSL Handshake Failed
Cloudflare could not negotiate a SSL/TLS handshake with the origin server."

aldavidson commented:
525 is OK, and clients that don't recognise it specifically will still get that it's a 5XX (it's not you, it's me) so that communicates the intent.

I'm still getting the same class-scope error, unfortunately, ebut in a different place now:

<img width="639" alt="screen shot 2017-10-13 at 10 12 27" src="https://user-images.githubusercontent.com/134501/31539115-89fed900-afff-11e7-817a-5117ee52364a.png">

aldavidson commented:
I had some more suggestions - I hope you don't mind, but I went ahead and refactored it a bit as explaining in github comments would have taken forever :)

I pushed it to a different branch: have a look at https://github.com/ministryofjustice/noms-api-gateway-management/pull/58 and see what you think

EdwardAndress commented:
That's a little embarrassing.  Anyway... done!

EdwardAndress commented:
@aldavidson - no problem at all.  Sounds great, thanks!  I will take a look now.
